### ISSUE TITLE
openssl pkcs11 engine: engine can't be loaded again after use and unload

### ISSUE THREAD

**Issue by sbrandt-aixtrusion** [OpenSC/libp11#556]
We have the following issue:
We need to unload the openssl pkcs11 engine after using it, to clear the internal EVP_PKEY cache. But, after unloading correctly(?), openssl is no longer able to find/load the engine again.

code flow:
* ENGINE_by_id("pkcs11")
* ENGINE_init, ctrl_cmd
* ENGINE_load_private_key
* ENGINE_finish
* ENGINE_free
* ENGINE_remove
* ENGINE_by_id("pkcs11") -> returns nullptr

```
4047A4C6277F0000:error:1280006A:DSO support routines:dlfcn_bind_func:could not bind to the requested symbol name:../crypto/dso/dso_dlfcn.c:188:symname(EVP_PKEY_base_id): /usr/lib/x86_64-linux-gnu/engines-3/pkcs11.so: undefined symbol: EVP_PKEY_base_id
4047A4C6277F0000:error:1280006A:DSO support routines:DSO_bind_func:could not bind to the requested symbol name:../crypto/dso/dso_lib.c:176:
4047A4C6277F0000:error:1300006D:engine routines:dynamic_load:init failed:../crypto/engine/eng_dyn.c:503:
4047A4C6277F0000:error:13000074:engine routines:ENGINE_by_id:no such engine:../crypto/engine/eng_list.c:430:id=pkcs11
```
About `EVP_PKEY_base_id` error, see below (OP-TEE version), probably not the actual problem.

Version of libraries used:
* x64 version
  * Ubuntu 22.04.05 LTS
  * OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)
  * openssl/jammy-updates,jammy-security,now 3.0.2-0ubuntu1.18 amd64
  * libsofthsm2/jammy,now 2.6.1-2ubuntu1 amd64
  * libp11-3/jammy-updates,now 0.4.11-1ubuntu0.22.04.1 amd64
* aarch64 version
  * Yocto kirkstone
  * openssl 3.0.15
  * libp11 0.4.12

Note that unloading and reloading work if the engine is not used, i.e., ENGINE_load_private_key not called.

The same error also appears if the engine is not used, and ENGINE_finish() is not called before _remove and _free.

This ticket is derived from #541 (original problem report) and #552 where the recommendation for reloading the engine was given.

Sample code. Note that a key pair with label `testkey` is required for the code to work; PIN is 12345.
```cpp
#include <stdexcept>

#include <stdio.h>

#include <openssl/evp.h>
#include <openssl/engine.h>
#include <openssl/err.h>

namespace
{
    ENGINE* loadEngine()
    {
        ENGINE_load_builtin_engines();
        auto eng = ENGINE_by_id("pkcs11");
        printf("ENGINE_by_id('pkcs11') -> %p\n", (void*)eng);
        if (!eng)
        {
            throw std::runtime_error("ENGINE_by_id('pkcs11') failed");
        }
        auto ret = ENGINE_init(eng);
        printf("ENGINE_init(%p) -> %d\n", (void*)eng, ret);
#if defined(__x86_64__)
        ret = ENGINE_ctrl_cmd(eng, "MODULE_PATH", 0, (void *)"/usr/lib/softhsm/libsofthsm2.so", NULL, 1);
#else
        ret = ENGINE_ctrl_cmd(eng, "MODULE_PATH", 0, (void *)"/usr/lib/libckteec.so.0", NULL, 1);
#endif
        printf("ENGINE_ctrl_cmd(%p, MODULE_PATH) -> %d\n", (void*)eng, ret);
        ret = ENGINE_ctrl_cmd_string(eng, "PIN", "12345", 0);
        printf("ENGINE_ctrl_cmd_string(%p, PIN) -> %d\n", (void*)eng, ret);
        return eng;
    }

    void loadAndFreePrivateKey(ENGINE* eng)
    {
        auto* private_key = ENGINE_load_private_key(eng, "pkcs11:object=testkey", NULL, NULL);
        printf("ENGINE_load_private_key() -> %p\n", (void*)private_key);
        EVP_PKEY_free(private_key);
        printf("EVP_PKEY_free(%p)\n", (void*)private_key);
    }

    void unloadEngine(ENGINE* eng)
    {
        auto ret = ENGINE_finish(eng);
        printf("ENGINE_finish(%p) -> %d\n", (void*)eng, ret);
        ret = ENGINE_free(eng);
        printf("ENGINE_free(%p) -> %d\n", (void*)eng, ret);
        ret = ENGINE_remove(eng);
        printf("ENGINE_remove(%p) -> %d\n", (void*)eng, ret);
    }

}

int main(int, char **)
{
    try
    {
        for (int i = 0; i < 3; i++)
        {
            ERR_clear_error();
            ENGINE* eng = loadEngine();
            loadAndFreePrivateKey(eng);
            unloadEngine(eng);
        }
        return 0;
    }
    catch (const std::exception& ex)
    {
        printf("Caught exception: %s\n", ex.what());
        ERR_print_errors_fp(stderr);
        return 1;
    }
}

```

Output of sample code:
```
ENGINE_by_id('pkcs11') -> 0x55bcac191a90
ENGINE_init(0x55bcac191a90) -> 1
ENGINE_ctrl_cmd(0x55bcac191a90, MODULE_PATH) -> 1
ENGINE_ctrl_cmd_string(0x55bcac191a90, PIN) -> 1
Found uninitialized token
Found uninitialized token
ENGINE_load_private_key() -> 0x55bcac1a5730
EVP_PKEY_free(0x55bcac1a5730)
ENGINE_finish(0x55bcac191a90) -> 1
ENGINE_free(0x55bcac191a90) -> 1
ENGINE_remove(0x55bcac191a90) -> 1
ENGINE_by_id('pkcs11') -> (nil)
Caught exception: ENGINE_by_id('pkcs11') failed
4047A4C6277F0000:error:1280006A:DSO support routines:dlfcn_bind_func:could not bind to the requested symbol name:../crypto/dso/dso_dlfcn.c:188:symname(EVP_PKEY_base_id): /usr/lib/x86_64-linux-gnu/engines-3/pkcs11.so: undefined symbol: EVP_PKEY_base_id
4047A4C6277F0000:error:1280006A:DSO support routines:DSO_bind_func:could not bind to the requested symbol name:../crypto/dso/dso_lib.c:176:
4047A4C6277F0000:error:1300006D:engine routines:dynamic_load:init failed:../crypto/engine/eng_dyn.c:503:
4047A4C6277F0000:error:13000074:engine routines:ENGINE_by_id:no such engine:../crypto/engine/eng_list.c:430:id=pkcs11
```

```
$ pkcs11-tool --list-objects --module /usr/lib/softhsm/libsofthsm2.so
Using slot 0 with a present token (0x9bdee03)
Public Key Object; EC  EC_POINT 384 bits
  EC_POINT:   ....
  EC_PARAMS:  06052b81040022
  label:      testkey
  ID:         1150
  Usage:      encrypt, verify, wrap
  Access:     local
```

Note that the error message from openssl changes when running on aarch64 with OP-TEE pkcs11 module. This is a hint that the `EVP_PKEY_base_id` error above is _not_ the actual problem:

```
ENGINE_by_id('pkcs11') -> 0xaaab07be8a10
ENGINE_init(0xaaab07be8a10) -> 1
ENGINE_ctrl_cmd(0xaaab07be8a10, MODULE_PATH) -> 1
ENGINE_ctrl_cmd_string(0xaaab07be8a10, PIN) -> 1
ENGINE_free(0xaaab07be8a10) -> 1
ENGINE_remove(0xaaab07be8a10) -> 1
ENGINE_by_id('pkcs11') -> (nil)
Caught exception: ENGINE_by_id('pkcs11') failed
40BF18BCFFFF0000:error:1300006D:engine routines:dynamic_load:init failed:../openssl-3.0.15/crypto/engine/eng_dyn.c:514:
40BF18BCFFFF0000:error:13000074:engine routines:ENGINE_by_id:no such engine:../openssl-3.0.15/crypto/engine/eng_list.c:433:id=pkcs11
```

**Comment by sbrandt-aixtrusion** [comment]
Seems that `engine_add_dynamic_id` refuses to re-add the engine as it is still in the `engine_dyn_list_head`.

**Comment by olszomal** [comment]
This issue has been resolved in PR#550.
@sbrandt-aixtrusion Could you please test again using pkcs11.so built from the latest master branch of https://github.com/OpenSC/libp11?

**Comment by sbrandt-aixtrusion** [comment]
With a simple test it works.

Unfortunately, the unloading now seems to break other pkcs11 stuff.
After unloading, we try to call `ckFunctions->C_FindObjectsInit(...);` which should be completely independent of the openssl engine usage, and we get `CKR_CRYPTOKI_NOT_INITIALIZED		(0x190UL)`.

So, we have to re-initialize our own pkcs11 session, which is probably possible, but I wonder if it is correct for openssl engine to completely deinit all pkcs11 sessions, and not just its own one?

Also, in the context of our main application, re-initializing the pkcs11 session does not yet work correctly ... will have to check there.

**Comment by mtrojnar** [comment]
Please open a separate issue when you have a demonstration code for that.  It looks to me like your PKCS#11 module fails to correctly handle interleaving `C_Initialize`/`C_Finalize` calls, but it is hard to be sure without a pkcs11-spy log.  Please make sure to include the log with the new issue. 

See https://www.cryptsoft.com/pkcs11doc/v210/group__SEC__11__4__GENERAL__PURPOSE__FUNCTIONS.html for the expected behavior of a PKCS#11 module:
>If several applications are using Cryptoki, each one should call [C_Finalize](https://www.cryptsoft.com/pkcs11doc/v210/pkcs11__all_8h.html#aC_Finalize). Each application's call to [C_Finalize](https://www.cryptsoft.com/pkcs11doc/v210/pkcs11__all_8h.html#aC_Finalize) should be preceded by a single call to [C_Initialize](https://www.cryptsoft.com/pkcs11doc/v210/pkcs11__all_8h.html#aC_Initialize) ; in between the two calls, an application can make calls to other Cryptoki functions. See [Section 6.5](https://www.cryptsoft.com/pkcs11doc/v210/group__SEC__6__5__APPLICATIONS__AND__THEIR__USE__OF__CRYPTOKI.html#SECTION_6_5) for more details.
> ''Despite the fact that the parameters supplied to [C_Initialize](https://www.cryptsoft.com/pkcs11doc/v210/pkcs11__all_8h.html#aC_Initialize) can in general allow for safe multi-threaded access to a Cryptoki library, the behavior of [C_Finalize](https://www.cryptsoft.com/pkcs11doc/v210/pkcs11__all_8h.html#aC_Finalize) is nevertheless undefined if it is called by an application while other threads of the application are making Cryptoki calls. The exception to this exceptional behavior of [C_Finalize](https://www.cryptsoft.com/pkcs11doc/v210/pkcs11__all_8h.html#aC_Finalize) occurs when a thread calls [C_Finalize](https://www.cryptsoft.com/pkcs11doc/v210/pkcs11__all_8h.html#aC_Finalize) while another of the application's threads is blocking on Cryptoki's [C_WaitForSlotEvent](https://www.cryptsoft.com/pkcs11doc/v210/pkcs11__all_8h.html#aC_WaitForSlotEvent) function. When this happens, the blocked thread becomes unblocked and returns the value [CKR_CRYPTOKI_NOT_INITIALIZED](https://www.cryptsoft.com/pkcs11doc/v210/pkcs11__all_8h.html#aCKR_CRYPTOKI_NOT_INITIALIZED)[.@see](mailto:.@see) [C_WaitForSlotEvent](https://www.cryptsoft.com/pkcs11doc/v210/pkcs11__all_8h.html#aC_WaitForSlotEvent) for more information.''
